src/sage/modules/free_module_pseudohomspace.py

```diff
@@ -0,0 +1,262 @@
+"""
+Space of pseudomorphisms of free modules
+
+AUTHORS:
+
+- Xavier Caruso, Yossef Musleh (2024-09): initial version
+"""
+# ****************************************************************************
+#  Copyright (C) 2024 Xavier Caruso <xavier.caruso@normalesup.org>
+#                     Yossef Musleh <specialholonomy@gmail.com>
+#
+#  Distributed under the terms of the GNU General Public License (GPL)
+#
+#    This code is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty
+#    of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+#  See the GNU General Public License for more details; the full text
+#  is available at:
+#
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+from sage.structure.unique_representation import UniqueRepresentation
+
+from sage.categories.homset import HomsetWithBase
+from sage.matrix.matrix_space import MatrixSpace
+from sage.structure.sequence import Sequence
+from sage.rings.polynomial.ore_polynomial_ring import OrePolynomialRing
+from sage.modules.free_module_pseudomorphism import FreeModulePseudoMorphism
+
+
+class FreeModulePseudoHomspace(UniqueRepresentation, HomsetWithBase):
+    r"""
+    This class implements the space of Pseudomorphisms with a fixed twist.
+
+    For free modules, the elements of a pseudomorphism correspond to matrices
+    which define the mapping on elements of a basis.
+
+    TESTS::
+
+        sage: F = GF(125)
+        sage: M = F^2
+        sage: Frob = F.frobenius_endomorphism()
+        sage: PHS = M.pseudoHom(Frob)
+        sage: h = PHS([[1, 2], [1, 1]])
+        sage: e = M((4*F.gen()^2 + F.gen() + 2, 4*F.gen()^2 + 4*F.gen() + 4))
+        sage: h(e)
+        (z3, 2*z3^2 + 3*z3 + 3)
+    """
+    Element = FreeModulePseudoMorphism
+
+    @staticmethod
+    def __classcall_private__(cls, domain, codomain, twist):
+        r"""
+        Constructs the space of pseudomorphisms with a given twist.
+
+        INPUT:
+
+        -  ``domain`` -- a free module,  the domain of this pseudomorphism
+
+        -  ``codomain`` -- a free module, the codomain of this pseudomorphism
+
+        -  ``twist`` -- a twisting morphism/derivation or a Ore polynomial ring
+
+        TESTS::
+
+            sage: F = GF(125)
+            sage: Frob = F.frobenius_endomorphism()
+            sage: M = F^2
+            sage: H = M.pseudoHom(Frob)
+            sage: type(H)
+            <class 'sage.modules.free_module_pseudohomspace.FreeModulePseudoHomspace_with_category'>
+
+            sage: # Testsuite(H).run()
+        """
```

Comment:
What is this commented out?

Reply:
It's because several tests (`_test_additive_associativity`, `_test_distributivity`, `_test_elements`, `_test_one`, `_test_zero`) fail :-). The point is that these tests do not actually make sense when the derivation is not zero: the zero morphism is not pseudo-linear, *etc.* In order words, this homset is not a module.

I don't know what is the best way to fix this.
One option would be to add a keyword `skip=...`; another one to redefine the tests in the class.
Another possibility I was thinking at while writing the ticket was to remove the twisting derivation from the parent, *i.e.* the parent only remembers the domain, the codomain and the twisting morphism. This allows to recover the module structure, I guess (but when adding two pseudo-morphisms, the derivation changes). However, I do not like separating the twisting morphism and the derivation because there are always together elsewhere in sage (especially in the implementation of Ore polynomials).

Reply:
What stops you from just adding `skip=...`?

Reply:
OK. It smells like we need a new category, but it is something to consider later.